KOGITO-7311: Mocking HTTP CloudEvents sink with Wiremock (#55)
kiegroup/kogito-docs: serverlessworkflow/modules/ROOT/pages/testing-and-troubleshooting/mocking-http-cloudevents-with-wiremock.adoc

= Mocking HTTP CloudEvents Sink with Wiremock
==========================================================
:compat-mode!:
// Metadata:
:description: Mocking HTTP CloudEvents sink with Wiremock
:keywords: kogito, workflow, quarkus, serverless, test, integration, wiremock, CloudEvents
// Referenced documentation pages.

> :basic_integration_test_with_restassured_guide: xref:testing-and-troubleshooting/basic-integration-tests-with-restassured.adoc

Reviewer: maybe append the "_url" suffix to these envs?

Author: I just appended the "_guide" for our guides references regarding is clear the rest of them are urls. I think is clear enough where the env vars point, but I can rename the external ones with the sufix "_url" even I think they are clear enough, and grouped in header making them easy to maintain. @ricardozanini,  WDYT?

Reviewer 2: I think we should take the same approach for every guide to have a clear pattern. IIRC we are using `_url` suffix for any URL variable reference, so we should stick to it. If we change, we need to change all.

Author: @spolti @ricardozanini I agree, Thanks both!
I've added the `_url` to the external references variables.

:recap_of_http_based_testing_in_jvm_mode: {quarkus_testing_guide}#recap-of-http-based-testing-in-jvm-mode
:awaitility: http://www.awaitility.org/
:awaitility_class: https://javadoc.io/static/org.awaitility/awaitility/4.2.0/org/awaitility/Awaitility.html
:wiremockserver_class: https://javadoc.io/doc/com.github.tomakehurst/wiremock/latest/com/github/tomakehurst/wiremock/WireMockServer.html
:wiremock: https://wiremock.org/
:wiremock_verifying: {wiremock}/docs/verifying/

This document describes how to test your Serverless Workflow application that uses HTTP {cloud_events_url}[*CloudEvents*] and {knative_sink_binding_impl}[*sink binding*].

The testing case followed in this document is based on the {kogito_sw_order_processing_example}[`serverless-workflow-order_processing`] example application.

[[sinkbinding-test-overview]]
== Overview of HTTP CloudEvents, SinkBinding and testing
--
The application we want to test must be configured to use {knative_eventing}[*Knative Eventing*].
Using standard HTTP POST requests to send and receive events between event producers and {knative_sink_binding_overview}[*sinks*].

> These events conform to the {cloud_events_url}[*CloudEvents*] specifications, which enables creating, parsing, sending, and receiving events in any programming language.

Reviewer 2: ```suggestion
These events conform to the {cloud_events_url}[*CloudEvents*] specification, which enables creating, parsing, sending, and receiving events in any programming language.
```

When you create an event source, you can specify a {knative_sink_binding_overview}[*sink*] where events are sent to, from the source.
A {knative_sink_binding_overview}[*sink*] is an Addressable or a Callable resource that can receive incoming events from other resources.

> Knative Services, Channels, and Brokers are all examples of sinks.

Reviewer 2: ```suggestion
Kubernetes Deployments, Services, Knative Services, Channels, and Brokers are all examples of sinks.
```

> In this guide we are covering the testing of the Knative Service configured as {knative_sink_binding_overview}[*sink*].

Reviewer 2: ```suggestion
This guide covers the testing of the Serverless Workflow application sending events to a Knative Service configured as {knative_sink_binding_overview}[*sink*].
```

The purpose of this guide is to mock that service to verify the cloud events are correctly received by the sink.
The {wiremock}[*Wiremok*] framework will add the mocked server that will verify the different Cloud Events received by the sink in the Serverless Workflow service execution.
--

[[mocking_sink_binding]]
== Testing the application with {knative_sink_binding_impl}[*SinkBinding*]

.Prerequisites

> * A completed Kogito Serverless Workflow application working. More information available in {getting_started_create_first_workflow_guide}[Creating your first Serverless Workflow service] guide.

Reviewer 2: I think it's useful also to link the events guide here (although to be published 😅 )

> * A completed Kogito Serverless Workflow application working. More information available in {getting_started_create_first_workflow_guide}[Creating your first Serverless Workflow service] guide.

Reviewer: ```suggestion
* A running Kogito Serverless Workflow application. More information available in {getting_started_create_first_workflow_guide}[Creating your first Serverless Workflow service] guide.
```

> * Enable {event_driven_architecture}[event-driven architecture] in the application using {knative_eventing}[*Knative Eventing*]

Reviewer 2: I think this item is unnecessary. What did you mean here?

> * Configure the application to use HTTP CloudEvents with {knative_sink_binding_impl}[*SinkBinding*].

Reviewer 2: ```suggestion
* Configure the workflow application to use HTTP CloudEvents with {knative_sink_binding_impl}[*SinkBinding*].
```

This may confuse users. Is this the application we are testing or the target one? The target (sink) doesn't matter since we won't test it, right? 

In this case, we need to evolve this description a little bit more if want to keep it here. Maybe linking back to our docs about Knative Eventing?

> === Add testing dependencies

Reviewer: ```suggestion
=== Adding tests dependencies
```

> Add the needed test dependencies to Serverless Workflow application pom.xml

Reviewer: maybe it worth to mention Gradle and Quarkus CLI options as well.

** HTTP based Testing in JVM mode dependencies. More detailed in {recap_of_http_based_testing_in_jvm_mode}[Testing your application] Quarkus guide.
** {wiremock}[Wiremock] framework. Allowing to mock the server that plays the Sink role.
** {awaitility}[Awaitility] used to express expectations of an asynchronous system.

.Application testing dependencies. See the example {kogito_sw_order_processing_example_pom}[`pom.xml`] for more details.
[source,xml]
----
<dependency>
    <groupId>io.quarkus</groupId>
    <artifactId>quarkus-junit5</artifactId>
    <scope>test</scope>
</dependency>
<dependency>
    <groupId>io.rest-assured</groupId>
    <artifactId>rest-assured</artifactId>
    <scope>test</scope>
</dependency>
<dependency>
  <groupId>com.github.tomakehurst</groupId>
  <artifactId>wiremock-jre8</artifactId>
  <scope>test</scope>
</dependency>
<dependency>
  <groupId>org.awaitility</groupId>
  <artifactId>awaitility</artifactId>
  <scope>test</scope>
</dependency>
----

=== Create the test class

> Create a test class mocking the Sink using Wiremock The follow piece of code explores the different elements that should compose it:

Reviewer 2: ```suggestion
Create a test class mocking the Sink using Wiremock. The following excerpt explores the different elements that should compose it:
```

        await() <11>
                .atMost(60, SECONDS)
                .with().pollInterval(1, SECONDS)
                .untilAsserted(() -> sink.verify (2, postRequestedFor(urlEqualTo("/")).withRequestBody(containing(order.getId())))); <12>
    }
}

----
<1> @QuarkusTest starts a Quarkus server for the whole lifetime of the tests execution run.
More details in {recap_of_http_based_testing_in_jvm_mode}[enable quarkus testing resources].
<2> The naming of the test would end with 'IT' to identify which test needs to be executed as an integration test.
<3> {wiremockserver_class}[WireMockServer] mocked server instance used for Sink Binding for testing.
<4> Used to test interactions with the application, more explained in the guide {basic_integration_test_with_restassured_guide}[Testing your Serverless Workflow application using REST Assured].
<5> `@BeforeAll` annotation used to signal that the annotated method should be executed before all the tests run.
<6> Creates the {wiremockserver_class}[WireMockServer] instance listening at the port passed as parameter and must match with the Sink configuration.
<7> Start the server before the tests executed.

> <8> Stub the mock API response.It accepts a MappingBuilder instance that we can use to build API mapping information such as URL, request parameters and body, headers, authorization etc.

Reviewer 3: ```suggestion
<8> Stub the mock API response. It accepts a MappingBuilder instance that we can use to build API mapping information such as URL, request parameters and body, headers, authorization etc.
```

<9> `@AfterAll` annotation used to signal that the annotated method should be executed after all the tests execution.
<10> Stop the server after all test execution.
<11> {awaitility_class}[`await()`] added to waiting for asynchronous operations.
<12> {wiremock_verifying}[`verify`] if the request has hit the mock API.

> 

Reviewer 2: Do you mind explaining what happens when we run this test? For example, the `await` part is super important, since it might shed a light into users' heads. We are waiting for the sink to receive the cloud event produced by the workflow. I think images would be nice here to illustrate it.

[NOTE]
====
It is very important to start the server before the tests execute, and stop the server after the tests finish.
We can reset the mock stubs in between the tests.
====

> === Configure the quarkus test

Reviewer 3: ```suggestion
=== Configure the Quarkus test
```

The {kogito_sw_order_processing_example}[example] is composed by three workflows:

.Example Order Processing workflows
image::testing-and-troubleshooting/order-example-worflows.png[]

The main workflow process the incoming Order event and start a parallel state calling two subflows: Fraud Handling and Shipping Handling.
The workflow will end once **both** subflows end.

Fraud Handling will produce a new `FraudEvaluation` event if the order is above 1000 USD.
Any other system or service in the architecture can then read this event and react upon it, like canceling the order for example.

In parallel, regarding or not the order would need fraud evaluation, the workflow will produce events classifying the required Shipping service: International or Domestic.
For this example, domestic shipping is any order with address within US.

The event flow between components in the example application is shown in this diagram

.Example event flow between components
image::testing-and-troubleshooting/example-components-interaction.png[]

Here, how the testing components replicates these relationships to verify the events received by the sink

.Example testing events flow.

> image::testing-and-troubleshooting/testing-eventing-sink.png[]]

Reviewer 2: I think in this image you can remove the "broker" since the events flow directly to the mock.
